// angular-infinite-scroll/fesm2015/angular-infinite-scroll.js
import { NgForOf, CommonModule } from '@angular/common';
import { BehaviorSubject, interval, merge, Observable, Subject } from 'rxjs';
import { exhaustMap, switchMap, take } from 'rxjs/operators';
import { Component, Input, IterableDiffers, NgZone, ViewChild, ViewContainerRef, ContentChild, ContentChildren, Directive, TemplateRef, NgModule } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/** @type {?} */
const DEFAULTS = {
    STEP: 4,
    POSITION: 0,
    OFFSET: 100,
    DELAY: 0
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * @abstract
 * @template T
 */
class InfiniteScroll {
    /**
     * @param {?} _differs
     * @param {?} zone
     */
    constructor(_differs, zone) {
        this._differs = _differs;
        this.zone = zone;
        this._userEnd$ = new Observable();
        this._end$ = new Subject();
        this._updateAfterRender$ = new Subject();
        this.position = DEFAULTS.POSITION;
        this.step = DEFAULTS.STEP;
        this.offset = DEFAULTS.OFFSET;
        this.delay = DEFAULTS.DELAY;
        this.loading$ = new BehaviorSubject(false);
        zone.runOutsideAngular(() => {
            window.addEventListener('scroll', () => this._update());
            window.addEventListener('resize', () => this._update());
        });
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this._subscriptionLoadingEnd = merge(this._userEnd$, this._updateAfterRender$).subscribe(() => this.loading$.next(false));
        this._subscriptionUpdateAfterRender =
            this._updateAfterRender$.pipe(switchMap(() => interval(this.delay).pipe(take(1)))).subscribe(() => this._update());
        this._update();
    }
    /**
     * @return {?}
     */
    ngDoCheck() {
        if (this._ngFor) {
            this._ngFor.ngDoCheck();
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.destroy(this._subscriptionEnd);
        this.destroy(this._subscriptionLoading);
        this.destroy(this._subscriptionLoadingEnd);
        this.destroy(this._subscriptionUpdateAfterRender);
    }
    /**
     * @param {?} viewContainer
     * @param {?} template
     * @return {?}
     */
    createNgFor(viewContainer, template) {
        this._ngFor = new NgForOf(viewContainer, template, this._differs);
    }
    /**
     * @param {?} loading
     * @return {?}
     */
    subscribeLoading(loading) {
        this.destroy(this._subscriptionLoading);
        this._subscriptionLoading = this.loading$.subscribe(loading);
    }
    /**
     * @param {?} scrollEnd
     * @return {?}
     */
    subscribeEnd(scrollEnd) {
        this.destroy(this._subscriptionEnd);
        this._userEnd$ = this._end$.pipe(exhaustMap(() => scrollEnd(this.position, this.step)));
        this._subscriptionEnd = this._userEnd$.subscribe(this.newItems.bind(this), () => this.newItems.bind(this)([]));
    }
    /**
     * @return {?}
     */
    _update() {
        /** @type {?} */
        const scrollHeight = window.innerHeight + window.scrollY;
        if (scrollHeight >= document.body.offsetHeight - this.offset) {
            this.zone.runOutsideAngular(() => {
                this.update();
            });
        }
    }
    /**
     * @param {?} subscription
     * @return {?}
     */
    destroy(subscription) {
        if (subscription && !subscription.closed) {
            subscription.unsubscribe();
        }
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * @template T
 */
class InfiniteStaticMarker {
    /**
     * @param {?} template
     */
    constructor(template) {
        this.template = template;
        this.enabled = false;
    }
}
InfiniteStaticMarker.decorators = [
    { type: Directive, args: [{ selector: '[infiniteStatic]' },] },
];
/** @nocollapse */
InfiniteStaticMarker.ctorParameters = () => [
    { type: TemplateRef }
];
/**
 * @template T
 */
class InfiniteTemplateMarker {
    /**
     * @param {?} template
     */
    constructor(template) {
        this.template = template;
    }
}
InfiniteTemplateMarker.decorators = [
    { type: Directive, args: [{ selector: '[infiniteTemplate]' },] },
];
/** @nocollapse */
InfiniteTemplateMarker.ctorParameters = () => [
    { type: TemplateRef }
];
/**
 * @template T
 */
class InfiniteScrollComponent extends InfiniteScroll {
    /**
     * @param {?} differs
     * @param {?} zone
     */
    constructor(differs, zone) {
        super(differs, zone);
        this._dummies = 0;
        this._outOfItems = false;
    }
    /**
     * @param {?} loading
     * @return {?}
     */
    set loading(loading) {
        this.subscribeLoading(loading);
    }
    /**
     * @param {?} scrollEnd
     * @return {?}
     */
    set end(scrollEnd) {
        this.subscribeEnd(scrollEnd);
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this._itemsStatic = this.staticMarkers.toArray();
        this.updateItems();
        this.update();
    }
    /**
     * @return {?}
     */
    update() {
        if (!this._items) {
            this._items = [];
            this._dummies = 0;
        }
        if (this._items && (!this._items.length || this._items.every((item) => item === undefined))) {
            this._items = [];
            this._dummies = 0;
            this.addDummies();
        }
        /** @type {?} */
        let staticLength = 0;
        if (this._itemsStatic) {
            staticLength = this._itemsStatic.length;
        }
        if (this.position < staticLength) {
            this.loading$.next(true);
            this._updateAfterRender$.next();
            this.updateItems();
            this.position += this.step;
        }
        else if (staticLength + this.position < this._items.length - this._dummies) {
            this.loading$.next(true);
            this._updateAfterRender$.next();
            this.position += this.step;
        }
        else if (this._subscriptionEnd) {
            this.loading$.next(true);
            this._end$.next();
            this.addDummies();
            this.position += this.step;
        }
    }
    /**
     * @param {?} newItems
     * @return {?}
     */
    newItems(newItems) {
        while (this._dummies > 0) {
            if (this._items.length) {
                this._items.pop();
            }
            this._dummies--;
        }
        /** @type {?} */
        const newItemsArray = Array.from(newItems);
        this.zone.run(() => {
            this._items = this._items.concat(newItemsArray);
        });
        // only continue when newItems arrive
        if (newItemsArray.length) {
            this._updateAfterRender$.next();
        }
        else {
            this._outOfItems = true;
        }
    }
    /**
     * @return {?}
     */
    updateItems() {
        this.zone.run(() => {
            for (const index in this._itemsStatic) {
                this._itemsStatic[index].enabled = this.position > index;
            }
        });
    }
    /**
     * @return {?}
     */
    addDummies() {
        this.zone.run(() => {
            if (!this._dummies && !this._outOfItems) {
                this._items = this._items.concat(Array(this.step).fill(undefined));
                this._dummies += this.step;
            }
        });
    }
}
InfiniteScrollComponent.decorators = [
    { type: Component, args: [{
                selector: 'infinite-scroll',
                template: `
    <ng-template ngFor let-item [ngForOf]="_itemsStatic">
      <ng-template [ngIf]="item.enabled">
        <ng-container *ngTemplateOutlet="item.template"></ng-container>
      </ng-template>
    </ng-template>
    <ng-template ngFor let-item [ngForOf]="_items">
      <ng-container *ngTemplateOutlet="templateMarker.template; context: {$implicit: item}"></ng-container>
    </ng-template>`
            },] },
];
/** @nocollapse */
InfiniteScrollComponent.ctorParameters = () => [
    { type: IterableDiffers },
    { type: NgZone }
];
InfiniteScrollComponent.propDecorators = {
    staticMarkers: [{ type: ContentChildren, args: [InfiniteStaticMarker,] }],
    templateMarker: [{ type: ContentChild, args: [InfiniteTemplateMarker,] }],
    dynamicTemplate: [{ type: ViewChild, args: ['dynamic',] }],
    position: [{ type: Input }],
    step: [{ type: Input }],
    offset: [{ type: Input }],
    delay: [{ type: Input }],
    loading: [{ type: Input }],
    end: [{ type: Input }]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * @template T
 */
class InfiniteScrollDirective extends InfiniteScroll {
    /**
     * @param {?} viewContainer
     * @param {?} templateRef
     * @param {?} differs
     * @param {?} zone
     */
    constructor(viewContainer, templateRef, differs, zone) {
        super(differs, zone);
        this._positionInitial = DEFAULTS.POSITION;
        this._dummies = 0;
        this._outOfItems = false;
        this.step = DEFAULTS.STEP;
        this.offset = DEFAULTS.OFFSET;
        this.delay = DEFAULTS.DELAY;
        this.createNgFor(viewContainer, templateRef);
    }
    /**
     * @param {?} infiniteScrollOf
     * @return {?}
     */
    set infiniteScrollOf(infiniteScrollOf) {
        if (infiniteScrollOf) {
            this._items = Array.from(infiniteScrollOf);
        }
        this.position = this._positionInitial;
        this._dummies = 0;
        this.updateItems();
        this.update();
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    set infiniteScrollTrackBy(fn) {
        this._ngFor.ngForTrackBy = fn;
    }
    /**
     * @return {?}
     */
    get infiniteScrollTrackBy() {
        return this._ngFor.ngForTrackBy;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set infiniteScrollTemplate(value) {
        this._ngFor.ngForTemplate = value;
    }
    /**
     * @param {?} position
     * @return {?}
     */
    set infiniteScrollPosition(position) {
        if (position === undefined || position === null) {
            this.position = DEFAULTS.POSITION;
        }
        else {
            this.position = position;
        }
        this._positionInitial = position;
    }
    /**
     * @param {?} loading
     * @return {?}
     */
    set infiniteScrollLoading(loading) {
        this.subscribeLoading(loading);
    }
    /**
     * @param {?} scrollEnd
     * @return {?}
     */
    set infiniteScrollEnd(scrollEnd) {
        this.subscribeEnd(scrollEnd);
    }
    /**
     * @return {?}
     */
    update() {
        if (!this._items) {
            this._items = [];
        }
        if (this._items && (!this._items.length || this._items.every((item) => item === undefined))) {
            this._items = [];
            this.addDummies();
            this.updateItems();
        }
        if (this.position < this._items.length - this._dummies) {
            this.loading$.next(true);
            this.updateItems();
            this._updateAfterRender$.next();
            this.position += this.step;
        }
        else if (this._subscriptionEnd) {
            this.loading$.next(true);
            this._end$.next();
            this.position += this.step;
            this.addDummies();
            this.updateItems();
        }
    }
    /**
     * @param {?} newItems
     * @return {?}
     */
    newItems(newItems) {
        while (this._dummies > 0) {
            if (this._items.length) {
                this._items.pop();
            }
            this._dummies--;
        }
        /** @type {?} */
        const newItemsArray = Array.from(newItems);
        this._items = this._items.concat(newItemsArray);
        this.updateItems();
        // only continue when newItems arrive
        if (newItemsArray.length) {
            this._updateAfterRender$.next();
        }
        else {
            this._outOfItems = true;
        }
    }
    /**
     * @return {?}
     */
    updateItems() {
        this.zone.run(() => {
            if (this._items) {
                // update ngForOf<T> directive
                this._ngFor.ngForOf = this._items.slice(0, this.position);
            }
        });
    }
    /**
     * @return {?}
     */
    addDummies() {
        if (!this._dummies && !this._outOfItems) {
            this._items = this._items.concat(Array(this.step).fill(undefined));
            this._dummies += this.step;
        }
    }
}
InfiniteScrollDirective.decorators = [
    { type: Directive, args: [{ selector: '[infiniteScroll]' },] },
];
/** @nocollapse */
InfiniteScrollDirective.ctorParameters = () => [
    { type: ViewContainerRef },
    { type: TemplateRef },
    { type: IterableDiffers },
    { type: NgZone }
];
InfiniteScrollDirective.propDecorators = {
    infiniteScrollOf: [{ type: Input }],
    infiniteScrollTrackBy: [{ type: Input }],
    infiniteScrollTemplate: [{ type: Input }],
    infiniteScrollPosition: [{ type: Input }],
    step: [{ type: Input, args: ['infiniteScrollStep',] }],
    offset: [{ type: Input, args: ['infiniteScrollOffset',] }],
    delay: [{ type: Input, args: ['infiniteScrollDelay',] }],
    infiniteScrollLoading: [{ type: Input }],
    infiniteScrollEnd: [{ type: Input }]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
class InfiniteScrollModule {
}
InfiniteScrollModule.decorators = [
    { type: NgModule, args: [{
                declarations: [InfiniteScrollDirective, InfiniteScrollComponent, InfiniteStaticMarker, InfiniteTemplateMarker],
                imports: [CommonModule],
                exports: [InfiniteScrollDirective, InfiniteScrollComponent, InfiniteStaticMarker, InfiniteTemplateMarker]
            },] },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

export { InfiniteScrollModule, InfiniteScroll as ɵb, InfiniteScrollComponent as ɵe, InfiniteStaticMarker as ɵc, InfiniteTemplateMarker as ɵd, InfiniteScrollDirective as ɵa };
