// angular-infinite-scroll/esm5/lib/infinite-scroll.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { NgForOf } from '@angular/common';
import { BehaviorSubject, interval, merge, Observable, Subject } from 'rxjs';
import { exhaustMap, switchMap, take } from 'rxjs/operators';
import { DEFAULTS } from './defaults';
/**
 * @abstract
 * @template T
 */
var /**
 * @abstract
 * @template T
 */
InfiniteScroll = /** @class */ (function () {
    function InfiniteScroll(_differs, zone) {
        var _this = this;
        this._differs = _differs;
        this.zone = zone;
        this._userEnd$ = new Observable();
        this._end$ = new Subject();
        this._updateAfterRender$ = new Subject();
        this.position = DEFAULTS.POSITION;
        this.step = DEFAULTS.STEP;
        this.offset = DEFAULTS.OFFSET;
        this.delay = DEFAULTS.DELAY;
        this.loading$ = new BehaviorSubject(false);
        zone.runOutsideAngular(function () {
            window.addEventListener('scroll', function () { return _this._update(); });
            window.addEventListener('resize', function () { return _this._update(); });
        });
    }
    /**
     * @return {?}
     */
    InfiniteScroll.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this._subscriptionLoadingEnd = merge(this._userEnd$, this._updateAfterRender$).subscribe(function () { return _this.loading$.next(false); });
        this._subscriptionUpdateAfterRender =
            this._updateAfterRender$.pipe(switchMap(function () { return interval(_this.delay).pipe(take(1)); })).subscribe(function () { return _this._update(); });
        this._update();
    };
    /**
     * @return {?}
     */
    InfiniteScroll.prototype.ngDoCheck = /**
     * @return {?}
     */
    function () {
        if (this._ngFor) {
            this._ngFor.ngDoCheck();
        }
    };
    /**
     * @return {?}
     */
    InfiniteScroll.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.destroy(this._subscriptionEnd);
        this.destroy(this._subscriptionLoading);
        this.destroy(this._subscriptionLoadingEnd);
        this.destroy(this._subscriptionUpdateAfterRender);
    };
    /**
     * @param {?} viewContainer
     * @param {?} template
     * @return {?}
     */
    InfiniteScroll.prototype.createNgFor = /**
     * @param {?} viewContainer
     * @param {?} template
     * @return {?}
     */
    function (viewContainer, template) {
        this._ngFor = new NgForOf(viewContainer, template, this._differs);
    };
    /**
     * @param {?} loading
     * @return {?}
     */
    InfiniteScroll.prototype.subscribeLoading = /**
     * @param {?} loading
     * @return {?}
     */
    function (loading) {
        this.destroy(this._subscriptionLoading);
        this._subscriptionLoading = this.loading$.subscribe(loading);
    };
    /**
     * @param {?} scrollEnd
     * @return {?}
     */
    InfiniteScroll.prototype.subscribeEnd = /**
     * @param {?} scrollEnd
     * @return {?}
     */
    function (scrollEnd) {
        var _this = this;
        this.destroy(this._subscriptionEnd);
        this._userEnd$ = this._end$.pipe(exhaustMap(function () { return scrollEnd(_this.position, _this.step); }));
        this._subscriptionEnd = this._userEnd$.subscribe(this.newItems.bind(this), function () { return _this.newItems.bind(_this)([]); });
    };
    /**
     * @return {?}
     */
    InfiniteScroll.prototype._update = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var scrollHeight = window.innerHeight + window.scrollY;
        if (scrollHeight >= document.body.offsetHeight - this.offset) {
            this.zone.runOutsideAngular(function () {
                _this.update();
            });
        }
    };
    /**
     * @param {?} subscription
     * @return {?}
     */
    InfiniteScroll.prototype.destroy = /**
     * @param {?} subscription
     * @return {?}
     */
    function (subscription) {
        if (subscription && !subscription.closed) {
            subscription.unsubscribe();
        }
    };
    return InfiniteScroll;
}());
/**
 * @abstract
 * @template T
 */
export { InfiniteScroll };
if (false) {
    /** @type {?} */
    InfiniteScroll.prototype._subscriptionLoading;
    /** @type {?} */
    InfiniteScroll.prototype._subscriptionLoadingEnd;
    /** @type {?} */
    InfiniteScroll.prototype._userEnd$;
    /** @type {?} */
    InfiniteScroll.prototype._subscriptionEnd;
    /** @type {?} */
    InfiniteScroll.prototype._subscriptionUpdateAfterRender;
    /** @type {?} */
    InfiniteScroll.prototype._end$;
    /** @type {?} */
    InfiniteScroll.prototype._updateAfterRender$;
    /** @type {?} */
    InfiniteScroll.prototype._ngFor;
    /** @type {?} */
    InfiniteScroll.prototype.position;
    /** @type {?} */
    InfiniteScroll.prototype.step;
    /** @type {?} */
    InfiniteScroll.prototype.offset;
    /** @type {?} */
    InfiniteScroll.prototype.delay;
    /** @type {?} */
    InfiniteScroll.prototype.loading$;
    /** @type {?} */
    InfiniteScroll.prototype._differs;
    /** @type {?} */
    InfiniteScroll.prototype.zone;
    /**
     * @abstract
     * @return {?}
     */
    InfiniteScroll.prototype.update = function () { };
    /**
     * @abstract
     * @param {?} newItems
     * @return {?}
     */
    InfiniteScroll.prototype.newItems = function (newItems) { };
}
